### PR TITLE
plan: forbid limit to push across selection.

### PR DESCRIPTION
**Pull request by hanfei1991**
When the SQL is `Select * from t left join s on true where least(1,2,3,t1.a,t2.b) > 0 order by t2.a limit 10`, the old plan is `Join(t,s)->TopN->Selection`, it's wrong. After fixing it, the plan is `Join(t,s)->Selection->TopN`.
@shenli @coocood @zimulala PTAL

### CODE REVIEW

File: plan/physical_plan_builder.go [hunk 1]
```diff
@@ -855,6 +866,8 @@ func (p *Selection) convert2PhysicalPlanEnforce(prop *requiredProperty) (*physic
 	if prop.limit != nil && len(prop.props) > 0 {
 		if t, ok := info.p.(physicalDistSQLPlan); ok {
 			t.addTopN(p.ctx, prop)
+		} else if _, ok := info.p.(*Selection); !ok {
```
Comment (coocood) [inline plan/physical_plan_builder.go]: If the child `info` is `*Selection`,  the Selection `p` will be lost?

Comment (hanfei1991) [inline plan/physical_plan_builder.go]: If child is selection, it means the selection has been processed

Comment (coocood) [inline plan/physical_plan_builder.go]: Oh, I remembered, the child has pulled the parent Selection down.
This design is anit-pattern.

File: plan/match_property.go [hunk 2]
```diff
@@ -231,10 +231,7 @@ func (p *Selection) matchProperty(prop *requiredProperty, childPlanInfo ...*phys
 		res.count = uint64(float64(res.count) * selectionFactor)
 		return res
 	}
-	np := *p
-	np.SetChildren(childPlanInfo[0].p)
-	count := uint64(float64(childPlanInfo[0].count) * selectionFactor)
-	return &physicalPlanInfo{p: &np, cost: childPlanInfo[0].cost, count: count}
+	return childPlanInfo[0]
```
Comment (coocood) [inline plan/match_property.go]: Here the `childPlanInfo` is actually not child, it is the selection itself? 
Then the `matchProperty` is not used as it is supposed to be.

Comment (hanfei1991) [inline plan/match_property.go]: Yeah, in case of not on table, it's the selection itself.

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**